homeassistant/components/onkyo/__init__.py

```diff
@@ -51,37 +52,47 @@ async def async_setup_entry(hass: HomeAssistant, entry: OnkyoConfigEntry) -> boo
 
     host = entry.data[CONF_HOST]
 
-    info = await async_interview(host)
+    try:
+        info = await async_interview(host)
+    except OSError as exc:
+        raise ConfigEntryNotReady(f"Unable to connect to: {host}") from exc
     if info is None:
         raise ConfigEntryNotReady(f"Unable to connect to: {host}")
 
-    receiver = await Receiver.async_create(info)
+    manager = ReceiverManager(hass, entry, info)
 
     sources_store: dict[str, str] = entry.options[OPTION_INPUT_SOURCES]
     sources = {InputSource(k): v for k, v in sources_store.items()}
 
     sound_modes_store: dict[str, str] = entry.options.get(OPTION_LISTENING_MODES, {})
     sound_modes = {ListeningMode(k): v for k, v in sound_modes_store.items()}
 
-    entry.runtime_data = OnkyoData(receiver, sources, sound_modes)
+    entry.runtime_data = OnkyoData(manager, sources, sound_modes)
 
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
-    await receiver.conn.connect()
+    manager_task = entry.async_create_background_task(
+        hass, manager.run(), "run_connection"
+    )
+    wait_for_started_task = asyncio.create_task(manager.started.wait())
+    done, _ = await asyncio.wait(
+        (manager_task, wait_for_started_task), return_when=asyncio.FIRST_COMPLETED
+    )
+    if manager_task in done:
+        # Something went wrong, so let's error out here by awaiting the task
+        try:
+            await manager_task
+        except OSError as exc:
+            raise ConfigEntryNotReady(f"Unable to connect to: {host}") from exc
 
     return True
 
 
 async def async_unload_entry(hass: HomeAssistant, entry: OnkyoConfigEntry) -> bool:
     """Unload Onkyo config entry."""
     del hass.data[DATA_MP_ENTITIES][entry.entry_id]
-
-    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
-
-    receiver = entry.runtime_data.receiver
-    receiver.conn.close()
-
-    return unload_ok
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    # the connection will be automatically closed when the background task is cancelled
```

Comment:
this comment seems out of place, as in, there's not code here that asks for explanation